enhancements/ebpf-in-ocp.md

```diff
@@ -0,0 +1,211 @@
+---
+title: ebpf-in-ocp
+authors:
+  - "@danwinship"
+reviewers:
+  - "@TBD, for network observability"
+  - "@TBD, for other existing eBPF use case X"
+  - "@TBD, for other future eBPF use case Y"
+  - "@dave-tucker, for bpfd"
+approvers:
+  - TBD
+api-approvers:
+  - None
+creation-date: 2022-05-25
+last-updated: 2022-05-25
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - TBD
+see-also:
+replaces:
+superseded-by:
+---
+
+# Guidelines for the Use of eBPF in OCP
+
+## Summary
+
+OCP developers and customers alike would like to use eBPF in OCP
+clusters. At the present time, _some_ eBPF hooks are designed to be
+"shareable", but others are not (and some are shareable only when used
+in certain restricted ways). We need to create rules for how OCP will
+use eBPF and how we expect others to use (or not use it) to prevent
+conflicts later.
+
+## Motivation
+
+### User Stories
+
+#### Existing Use Case Number 1
+
+(Something about observability or tracing where people are already
+using eBPF in OCP.)
+
+#### Upcoming Network Observability Features
+
+(Something about upcoming network observability features.)
+
+#### Ensuring Network (and General) Supportability
+
+As an OCP developer or support engineer, I would like to be able to
+debug customer problems without having to worry that random
+"unsanctioned" eBPF is messing things up in inexplicable ways.
+
+#### Customer eBPF Bug Patching
+
+As an OCP customer, I want to work around an OCP/RHEL bug by using a
+`MachineConfig` (or something similar) to deploy a small eBPF program
+to every node in my cluster, without voiding my warranty.
+
+I might be OK with needing a support exception, since this is just to
+work around a bug that should eventually be fixed.
+
+#### Customer eBPF Feature Patching
+
+As an OCP customer, I want to add functionality to OCP/RHEL by using a
+`MachineConfig` (or something similar) to deploy a small (or
+not-so-small) eBPF program to every node in my cluster, without
+voiding my warranty.
+
+I might be OK with needing a support exception.
+
+(eg, concretely, there is a customer who wants to use eBPF to
+implement [OVS-style SLB
+Bonding](https://docs.openvswitch.org/en/latest/topics/bonding/#slb-bonding)
+on a kernel bond interface, using eBPF. The upstream Linux maintainer
+rejected a patch to implement it in the kernel driver directly, in
+part because it would be so easy to implement it with eBPF.)
+
+#### Cilium as a Supported Third-Party Network Plugin
+
+As an OCP customer, I want to deploy an OCP cluster using Cilium as a
+certified third-party network plugin, without worrying that any other
+OCP components are going to interfere with it (or vice versa).
+
+Cilium uses a wide variety of eBPF hooks, including FIXME.
```

Comment:
This talk on Cilium's experience seems relevant: https://lpc.events/event/16/contributions/1353/attachments/1068/2130/plumbers_2022_tc_bpf_links.pdf

tl;dr - they saw datadog stomping on Cilium's hooks in a customer environment; they believe bpf_link attachment semantics help with this problem; they propose a model for using bpf_link with tc

Reply:
Yes, @markmc this example of (Datadog) BPF stomping on (Cilium) BPF is exactly the problem I'm trying to highlight.

For legal/support reasons RH must document what BPF hooks our products use and state RH cannot support shared usage of said BPF hooks. Regardless this stomping _will occur in practice_ in production environments, leading to **hard-to-debug** scenarios.

**IMHO** we should implement a technical solution, that:

1. as minimum detects and reports if a products BPF hook get "stomped"
2. preferably detects and _denies_ others from taking over products BPF hook